### PR TITLE
[libc][cmake] reset COMPILE_DEFINITIONS

### LINKED ISSUE #77395: [libc] fix -Wshorten-64-to-32 for 32b arm mmap
https://github.com/llvm/llvm-project/pull/77350 was an attempted fix that didn't work.  Filing a TODO to track a quick workaround to get the build bot green, until we can fix this properly.

### PR DESCRIPTION
**Pull request by nickdesaulniers**
While trying to enable -Werror (#74506), the 32b ARM build bot reported an
error stemming from -Wshorten-64-to-32 related to usages of `off_t`.

I failed to fix these properly in #77350 (the 32b ARM build is not a fullbuild)
and #77396.

It turns out, the preprocessor defines `-D_LARGEFILE_SOURCE` and
`-D_FILE_OFFSET_BITS=64` were being set for llvmlibc when using the cmake build
system. In particular, these preprocessor defines are feature test macros used
by glibc, and which have effects no the corresponding ABI for types like
`off_t` (for instance, should `off_t` be 32b or 64b on 32b targets).

But who was setting these?  Turns out that the use of add_compile_definitions
in llvm/cmake/modules/HandleLLVMOptions.cmake was setting these (and more),
which is then inherited by every subdirectory.  While some of these defines
maybe make sense for host builds, they do not make sense for libraries for the
target. The full list of defines being set prior to this commit:

- `-D_GNU_SOURCE`
- `-D_FILE_OFFSET_BITS=64`
- `-D_DEBUG`
- `-D_GLIBCXX_ASSERTIONS`
- `-D_LARGEFILE_SOURCE`
- `-D_FILE_OFFSET_BITS=64`
- `-D__STDC_CONSTANT_MACROS`
- `-D__STDC_FORMAT_MACROS`
- `-D__STDC_LIMIT_MACROS`

If we desire any of the above, we should manually reset them.

Fix this by resetting COMPILE_DEFINITIONS for the libc/ subdirectory.

Side note: to debug 'directory properties' in cmake, you first need to use
`get_directory_property` to fetch the corresponding value into a variable
first, then that variable can be printed via `message`.

Link: https://www.gnu.org/software/libc/manual/html_node/Feature-Test-Macros.html#index-_005fFILE_005fOFFSET_005fBITS
Link: https://www.gnu.org/software/libc/manual/html_node/Feature-Test-Macros.html#index-_005fFILE_005fOFFSET_005fBITS

Fixes: #77395

### CODE REVIEW

**Review by lntue — APPROVED** [review]

**Comment by nickdesaulniers** [comment]
oops, meant to remove the changes to mmap.{c|h}. Removed.  We only need changes to the cmake.

**Comment by nickdesaulniers** [comment]
(after this, I'll revert the two bad patches of mine, then restart the 32b arm buildbots)

**Comment by nickdesaulniers** [comment]
> I would guess that the debug macro is being set based on the cmake flag -DCMAKE_BUILD_TYPE=Debug, and we might want to respect that in our build. Other than that this looks fine to me.

I just tested that hypothesis by configuring with `-DCMAKE_BUILD_TYPE=Release` (and clearing the cmake cache). You're right; the list of preprocessor defines changes to:

- _GNU_SOURCE
- _FILE_OFFSET_BITS=64
- _LARGEFILE_SOURCE_FILE_OFFSET_BITS=64
- __STDC_CONSTANT_MACROS
- __STDC_FORMAT_MACROS
- __STDC_LIMIT_MACROS

So I think it makes sense to:
1. clear everything
2. re-set _DEBUG if necessary.

I'll add 2.

**Comment by lntue** [comment]
> > I would guess that the debug macro is being set based on the cmake flag -DCMAKE_BUILD_TYPE=Debug, and we might want to respect that in our build. Other than that this looks fine to me.
> 
> I just tested that hypothesis by configuring with `-DCMAKE_BUILD_TYPE=Release` (and clearing the cmake cache). You're right; the list of preprocessor defines changes to:
> 
> * _GNU_SOURCE
> * _FILE_OFFSET_BITS=64
> * _LARGEFILE_SOURCE_FILE_OFFSET_BITS=64
> * __STDC_CONSTANT_MACROS
> * __STDC_FORMAT_MACROS
> * __STDC_LIMIT_MACROS
> 
> So I think it makes sense to:
> 
> 1. clear everything
> 2. re-set _DEBUG if necessary.
> 
> I'll add 2.

In that case, since there might be more of these with different configurations and invocation, can you  printout `COMPILE_DEFINITIONS` before clearing it in verbose mode, and/or just remove these flags for now?

**Comment by nickdesaulniers** [comment]
> In that case, since there might be more of these with different configurations and invocation

In that case, I suspect YAGNI, but adding logging and re-added _DEBUG in https://github.com/llvm/llvm-project/pull/77810/commits/fdc1e357266e1e5fc8c2c72cf3ef13350d0ee79e, PTAL.

File: libc/CMakeLists.txt [hunk 1]
```diff
@@ -7,10 +7,24 @@ endif()
...
+set_directory_properties(PROPERTIES
+  # `llvm-project/llvm/CMakeLists.txt` adds the following directive
+  # `include_directories( ${LLVM_INCLUDE_DIR} ${LLVM_MAIN_INCLUDE_DIR})` We
+  # undo it to be able to precisely control what is getting included.
+  INCLUDE_DIRECTORIES ""
+  # `llvm/cmake/modules/HandleLLVMOptions.cmake` uses `add_compile_definitions`
+  # to set a few preprocessor defines which we do not want.
+  COMPILE_DEFINITIONS ""
+)
+if (CMAKE_BUILD_TYPE STREQUAL "Debug")
```
Comment (lntue) [inline libc/CMakeLists.txt:25]: Do you mind checking what else will be set for the other two official CMake build types: RelWithDebInfo and MinSizeRel?

**Review by michaelrj-google — APPROVED** [review]
LGTM

File: libc/CMakeLists.txt (hunk 1 above)
Comment (nickdesaulniers) [reply libc/CMakeLists.txt:25]: RelWithDebInfo (BEFORE): _GNU_SOURCE_FILE_OFFSET_BITS=64_LARGEFILE_SOURCE_FILE_OFFSET_BITS=64__STDC_CONSTANT_MACROS__STDC_FORMAT_MACROS__STDC_LIMIT_MACROS

(Nothing new)

MinSizeRel: (BEFORE):
_GNU_SOURCE_FILE_OFFSET_BITS=64_LARGEFILE_SOURCE_FILE_OFFSET_BITS=64__STDC_CONSTANT_MACROS__STDC_FORMAT_MACROS__STDC_LIMIT_MACROS

(Nothing new)

**Review by lntue — APPROVED** [review]

**Comment by ldionne** [comment]
I think this makes sense, however I'll note that other things can be polluted by the "rest" of LLVM as well:
- Link flags
- CMake variables like `CMAKE_POSITION_INDEPENDENT_CODE`
- Other compiler flags that are not `-D` options

This is better than nothing, but I think the full fix is to gradually refactor the CMake files used by runtimes to avoid setting anything global.

**Comment by nickdesaulniers** [comment]
> This is better than nothing, but I think the full fix is to gradually refactor the CMake files used by runtimes to avoid setting anything global.

:100:

**Comment by nickdesaulniers** [comment]
I've filed https://github.com/llvm/llvm-project/issues/78479 as a follow up to the suggestions here and from Discord.

**Pull request merged**